In [1]:
# The code was removed by Watson Studio for sharing.

# Clustering Neighbourhoods in Toronto

This notebook explores possible neighbourhood clusters in Toronto.

In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [3]:
# The code was removed by Watson Studio for sharing.

After importing the data frame from the previous exercise, we will filter for boroughs with Toronto in their name to limit the analysis to the city.

In [4]:
# select only observations that contain Toronto in the name of their Borough and check the result
my_df = df_data_1[df_data_1['Borough'].str.contains('Toronto')]
my_df.head(10)

PostalCode          Borough  \
37        M4E     East Toronto   
41        M4K     East Toronto   
42        M4L     East Toronto   
43        M4M     East Toronto   
44        M4N  Central Toronto   
45        M4P  Central Toronto   
46        M4R  Central Toronto   
47        M4S  Central Toronto   
48        M4T  Central Toronto   
49        M4V  Central Toronto   

                                        Neighbourhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     The Beaches West, India Bazaar  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                                 North Toronto West  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Deer Park, Forest Hill SE, Rathnelly, South Hi...  43.686412 -79.400049

Now that we formatted the data in the desire format, we'll visualize the neighbourhoods in Toronto.

In [5]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [6]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 11)

# add markers to map
for lat, lng, label in zip(my_df['Latitude'], my_df['Longitude'], my_df['Neighbourhood']):
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [7]:
# The code was removed by Watson Studio for sharing.

We will create a funcion that will allow us to explore all neighbourhoods in Toronto.

In [11]:
LIMIT = 100
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [12]:
toronto_venues = getNearbyVenues(names = my_df['Neighbourhood'],
                                   latitudes = my_df['Latitude'],
                                   longitudes = my_df['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

We imported information about venues in all neighbourhoods of interest. Let's check the size and first couple of rows of the resulting dataframe.

In [13]:
print(toronto_venues.shape)
toronto_venues.head()

(1695, 7)


Neighbourhood  Neighbourhood Latitude  \
0                   The Beaches               43.676357   
1                   The Beaches               43.676357   
2                   The Beaches               43.676357   
3                   The Beaches               43.676357   
4  The Danforth West, Riverdale               43.679557   

   Neighbourhood Longitude                               Venue  \
0               -79.293031  The Big Carrot Natural Food Market   
1               -79.293031                 Grover Pub and Grub   
2               -79.293031                   Glen Manor Ravine   
3               -79.293031                       Upper Beaches   
4               -79.352188                            Pantheon   

   Venue Latitude  Venue Longitude     Venue Category  
0       43.678879       -79.297734  Health Food Store  
1       43.679181       -79.297215                Pub  
2       43.676821       -79.293942              Trail  
3       43.680563       -79.292869       Neighborhood  
4       43.677621       -79.351434   Greek Restaurant

Let's check how many venues were returned for each neighborhood.

In [14]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            21   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             46   
Central Bay Street                                                      88   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                15   
Church and Wellesley                                                    87   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              38   
Davisville North                                                         8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                      14   
Design Exchange, Toronto Dominion Centre                               100   
Dovercourt Village, Dufferin                                            20   
First Canadian Place, Underground city                                 100   
Forest Hill North, Forest Hill West                                      4   
Harbord, University of Toronto                                          34   
Harbourfront East, Toronto Islands, Union Station                      100   
Harbourfront, Regent Park                                               48   
High Park, The Junction South                                           23   
Lawrence Park                                                            3   
Little Portugal, Trinity                                                62   
Moore Park, Summerhill East                                              2   
North Toronto West                                                      16   
Parkdale, Roncesvalles                                                  15   
Rosedale                                                                 4   
Roselawn                                                                 2   
Runnymede, Swansea                                                      38   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         94   
Studio District                                                         38   
The Annex, North Midtown, Yorkville                                     23   
The Beaches                                                              4   
The Beaches West, India Bazaar                                          18   
The Danforth West, Riverdale                                            42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide, King, Richmond                                                100   
Berczy Park                                                              55   
Brockton, Exhibition Place, Parkdale Village                             21   
Business Reply Mail Processing Centre 969 Eastern                        19   
CN Tower, Bathurst Quay, Island airport, Harbou...                       14   
Cabbagetown, St. James Town                                              46   
Central Bay Street                                                       88   
Chinatown, Grange Park, Kensington Market                               100   
Christie                                                                 15   
Church and Wellesley                                          

Let's find out how many unique categories can be curated from all the returned venues.

In [15]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


To analyze the neighbourhoos, we need to do one hot encodings for the venue categories, which is done in the next cell. The resulting data frame should have 237 (unique categories) + 1 (Neighbourhood) columns. 

In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.shape

(1695, 238)

Now group by Neighbourhood and take mean of the observations. Confirm the new size which should correspond to the number of neighbourhoods in the row dimension. 

In [30]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.shape

(38, 238)

Let's get 10 most common venues for each neighbourhood and put that into a data frame. Inspect the first couple of rows of the result.

In [31]:
num_top_venues = 10

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [32]:
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns = columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant            Steakhouse   
1          Cocktail Bar            Steakhouse                  Café   
2                  Café        Breakfast Spot           Yoga Studio   
3           Pizza Place               Brewery            Skate Park   
4       Airport Service      Airport Terminal              Boutique   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     American Restaurant                   Bar                Bakery   
1      Italian Restaurant        Farmers Market           Cheese Shop   
2  Furniture / Home Store     Convenience Store          Climbing Gym   
3           Burrito Place               Butcher            Restaurant   
4        Sculpture Garden                 Plane               Airport   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Cosmetics Shop          Burger Joint                  Hotel  
1    Seafood Restaurant              Beer Bar                 Bakery  
2            Restaurant  Caribbean Restaurant          Burrito Place  
3      Recording Studio            Comic Shop                    Spa  
4    Airport Food Court          Airport Gate        Harbor / Marina

Now we are ready to do the clustering. After experimenting with a couple of k values, I settled for k = 6, although there is no right answer as for how many clusters there should be. Moreover, this data doesn't seem to cluster well in general.

In [33]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters = kclusters, random_state = 0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:38]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 5, 1, 1, 1, 1, 4, 1, 1, 1, 1, 4,
       1, 3, 1, 1, 4, 2, 1, 1, 1, 1, 1, 1, 0, 1, 1], dtype=int32)

We will combine all relevant information into one data frame and confirm the new shape.

In [34]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = my_df

# merge the data frames to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on = 'Neighbourhood')

toronto_merged.head()

PostalCode          Borough                   Neighbourhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               0     Health Food Store                   Pub   
41 -79.352188               1      Greek Restaurant           Coffee Shop   
42 -79.315572               1                  Park                   Gym   
43 -79.340923               1                  Café           Coffee Shop   
44 -79.388790               4                  Park           Swim School   

   3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
37                 Trail          Neighborhood                 Dog Run   
41    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
42    Italian Restaurant           Pizza Place                     Pub   
43             Gastropub    Italian Restaurant                  Bakery   
44              Bus Line           Yoga Studio        Doner Restaurant   

   6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
37   Filipino Restaurant  Fast Food Restaurant           Farmers Market   
41           Yoga Studio               Brewery  Fruit & Vegetable Store   
42         Movie Theater        Sandwich Place            Burrito Place   
43   American Restaurant           Yoga Studio     Gym / Fitness Center   
44   Filipino Restaurant  Fast Food Restaurant           Farmers Market   

   9th Most Common Venue 10th Most Common Venue  
37    Falafel Restaurant            Event Space  
41             Juice Bar           Liquor Store  
42          Burger Joint                Brewery  
43               Brewery     Seafood Restaurant  
44    Falafel Restaurant            Event Space

In [35]:
toronto_merged.shape

(38, 16)

And finally, we will visualize the clusters on a Toronto map. Our suspicion is confirmed, the clustering algorithm basically created one big cluster of neighbourhoods in the center and western part of the town, the rest is very granular. 

In [36]:
# create map
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We can now examine each cluster in more detail and look what specific venue categories distinguish each cluster. Probably the major thing that distinguishes the biggest cluster from the rest is that its neighbourhoods contain Coffee Shops and Cafes between top catetories. These two categories are basically missing among the top categories in the rest of the clusters.

In [47]:
print("Cluster 1:")
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Cluster 1:


Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
37   The Beaches     Health Food Store                   Pub   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37                 Trail          Neighborhood               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37   Filipino Restaurant  Fast Food Restaurant        Farmers Market   

   9th Most Common Venue 10th Most Common Venue  
37    Falafel Restaurant            Event Space

In [48]:
print("Cluster 2:")
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Cluster 2:


Neighbourhood 1st Most Common Venue  \
41                       The Danforth West, Riverdale      Greek Restaurant   
42                     The Beaches West, India Bazaar                  Park   
43                                    Studio District                  Café   
46                                 North Toronto West           Coffee Shop   
47                                         Davisville           Pizza Place   
49  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
51                        Cabbagetown, St. James Town           Coffee Shop   
52                               Church and Wellesley   Japanese Restaurant   
53                          Harbourfront, Regent Park           Coffee Shop   
54                           Ryerson, Garden District           Coffee Shop   
55                                     St. James Town           Coffee Shop   
56                                        Berczy Park           Coffee Shop   
57                                 Central Bay Street           Coffee Shop   
58                           Adelaide, King, Richmond           Coffee Shop   
59  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
60           Design Exchange, Toronto Dominion Centre           Coffee Shop   
61                     Commerce Court, Victoria Hotel           Coffee Shop   
65                The Annex, North Midtown, Yorkville           Coffee Shop   
66                     Harbord, University of Toronto                  Café   
67          Chinatown, Grange Park, Kensington Market                  Café   
68  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
69                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
70             First Canadian Place, Underground city                  Café   
75                                           Christie         Grocery Store   
76                       Dovercourt Village, Dufferin           Supermarket   
77                           Little Portugal, Trinity                   Bar   
78       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
82                      High Park, The Junction South    Mexican Restaurant   
83                             Parkdale, Roncesvalles        Breakfast Spot   
84                                 Runnymede, Swansea           Pizza Place   
87  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
41                    Coffee Shop    Italian Restaurant        Ice Cream Shop   
42                            Gym    Italian Restaurant           Pizza Place   
43                    Coffee Shop             Gastropub    Italian Restaurant   
46                    Yoga Studio            Bagel Shop                  Park   
47                   Dessert Shop        Sandwich Place    Italian Restaurant   
49                            Pub          Liquor Store      Sushi Restaurant   
51                     Restaurant           Pizza Place                  Café   
52                    Coffee Shop      Sushi Restaurant            Restaurant   
53                           Park                Bakery                   Pub   
54                 Clothing Store        Cosmetics Shop                  Café   
55                           Café                 Hotel            Restaurant   
56                   Cocktail Bar            Steakhouse                  Café   
57             Italian Restaurant                  Café        Sandwich Place   
58                           Café       Thai Restaurant            Steakhouse   
59                       Aquarium                 Hotel    Italian Restaurant   
60                           Café                 Hotel            Restaurant   
61                           Café                 Hotel            Restaurant   
65                           Café        Sandwich Pla

In [49]:
print("Cluster 3:")
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Cluster 3:


Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
63      Roselawn        Ice Cream Shop                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
63           Yoga Studio      Doner Restaurant     Fish & Chips Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
63   Filipino Restaurant  Fast Food Restaurant        Farmers Market   

   9th Most Common Venue 10th Most Common Venue  
63    Falafel Restaurant            Event Space

In [50]:
print("Cluster 4:")
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Cluster 4:


Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
48  Moore Park, Summerhill East            Playground          Tennis Court   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
48          Concert Hall     Convenience Store   Filipino Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
48  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   9th Most Common Venue 10th Most Common Venue  
48           Event Space   Ethiopian Restaurant

In [51]:
print("Cluster 5:")
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Cluster 5:


Neighbourhood 1st Most Common Venue  \
44                        Lawrence Park                  Park   
50                             Rosedale                  Park   
64  Forest Hill North, Forest Hill West      Sushi Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
44           Swim School              Bus Line           Yoga Studio   
50            Playground                 Trail               Dog Run   
64                  Park         Jewelry Store                 Trail   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
44      Doner Restaurant   Filipino Restaurant  Fast Food Restaurant   
50     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
64           Yoga Studio     Electronics Store      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
44        Farmers Market    Falafel Restaurant                  Event Space  
50        Farmers Market    Falafel Restaurant                  Event Space  
64            Donut Shop   Dumpling Restaurant  Eastern European Restaurant

In [52]:
print("Cluster 6:")
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Cluster 6:


Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
45  Davisville North                  Park        Clothing Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45        Breakfast Spot                   Gym                 Hotel   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45         Grocery Store     Food & Drink Shop        Sandwich Place   

   9th Most Common Venue 10th Most Common Venue  
45            Donut Shop    Dumpling Restaurant